### ISSUE TITLE
How to use this library to fetch HTML web page?

### ISSUE THREAD

**Issue by vfonic** [socketry/async-rest#2]
Hey Sam!

I'm trying to use this gem to build an API/HTML crawler. Currently I'm trying to fetch a single HTML web page and for some reason it's failing. I tried couple of different websites and got two outputs (below). Here's the code:

```ruby
require 'async/rest'

URL = 'https://xkcd.com/2205/'
Async::REST::Resource.for(URL) do |resource|
  query = resource.get

  pp "query.metadata: #{query.metadata.inspect}"
  pp "query.value: #{query.value.inspect}"
end
```

Outputs:

```ruby
 > ruby asink.rb
 0.41s    error: <Async::Task:0x3fecc68c263c connected to #<Addrinfo: 151.101.192.67:443 TCP (xkcd.com)> [fd=16] failed> [pid=31826] [2019-09-25 11:29:13 +0200]
               |   Async::REST::Error: Unknown content type: text/html!
               |   → /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/wrapper/json.rb:75 in `process_response'
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/representation.rb:67 in `process_response'
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/representation.rb:80 in `block (2 levels) in <class
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/representation.rb:92 in `value!'
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/representation.rb:107 in `value'
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/resource.rb:84 in `tap'
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/resource.rb:84 in `get'
               |     asink.rb:186 in `block in <main>'
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-rest-0.10.1/lib/async/rest/resource.rb:63 in `block in for'
               |     /Users/viktor/.rvm/gems/ruby-2.6.3/gems/async-1.21.0/lib/async/task.rb:248 in `block in make_fiber'
```

```ruby
> ruby asink.rb
"query.metadata: {}"
"query.value: nil"
 1.21s     warn: #<Async::HTTP::Client:0x00007f9d7ca65c90> [pid=31792] [2019-09-25 11:27:46 +0200]
               | Waiting for pool to drain: #<Async::HTTP::Pool resources=1/1/1;1/1/1 limit=nil>
```

The second output seems to be successful, but it seems like maybe it's a bug when the response body is empty in that it blocks, waiting for the data to be read...I think.

I also played around with `async` and `async-await` and managed to successfully use `async-http`'s `Async::HTTP::Internet` to get about 6x speed-up when fetching ~50 xkcd pages (thank you!!)
I haven't tried `async-io`. But all in all, I'm looking for an async HTTP fetching.

One thing I noticed is that the pages are always output in the same order, and it's still about 4x slower than fetching the same pages on node.js v12. I'm using ruby 2.6.3. I remember you were telling me that fibers will be faster in newer ruby versions. Is that the case or am I doing something wrong?

Thanks!

**Comment by ioquatix** [comment]
This gem is kind of annoying, even to me. Because, I followed REST thesis. But almost no web application follows REST thesis, it's mostly just a bunch of random JSON and other crap. REST thesis is very specific.

I'll review your specific case. However, there is nothing wrong with just using `async-http` to pull down HTML.

That being said, async-rest is good. Like, if you agree with REST thesis, `async-rest` reproduces it in the most logical way possible.

The way to pull HTML with `async-rest` is by ensuring you explain to `async-rest` what to do with it. For example, your model might just be a string, or it might be a nokogiri document.

That's why it's telling you "Unknown content type". Because by default it only understands JSON. Maybe we should build in support for HTML too, but it's a bit tricky because there are so many ways to "represent" an HTML document (e.g. Nokogiri, String, some other DOM).

**Comment by ioquatix** [comment]
By the way, things like `RestClient` are not following REST thesis very well. The closest thing they have is `RestClient::Resource`. A pessimistic opinion of `RestClient` is that it really has very little to do with the REST thesis. So, I do understand it's super confusing, even to me before I studied the thesis.

**Comment by vfonic** [comment]
If you decide to make HTML built in, I'd suggest to simply treat it as String...in which case it's not really following REST thesis anymore, but it's just a bunch of random string and other crap. :)

I'll see what I can do with `async-http`. It looks promising and easy to use. I saw this comment in the `async-http` [README.md](https://github.com/socketry/async-http#post-json-data) and thought maybe `async-rest` is always preferred:
> Consider using async-rest instead.

Re-reading it now, it seems it's only suggested for JSON resources.

I'll close this and open a new issue if I run into any problems using `async-http`.

**Comment by ioquatix** [comment]
For interfacing with programmatic interfaces, `async-rest` can be great. Otherwise, for just fetching data, images, whatever, use straight `async-http`.

**Comment by vfonic** [comment]
I'm currently crawling a website that has no API, so I'm fetching HTML. In this case, I'll stick with `async-http`.

Thanks!

**Comment by ioquatix** [comment]
I hacked together a small example just for interest.

**Comment by vfonic** [comment]
F*ckin awesome! :D

So if I wanted to use this asynchronously to fetch multiple comics, which part is actually async?
`Async do` block or `Async::REST::Resource.for(URL)` or both or?

https://github.com/socketry/async-rest/commit/4580029adac0cb655745ad438b7563ccef12b4bc#diff-6e27a3c87d3d56d60333b3dfef8e12d4R52-R59

**Comment by ioquatix** [comment]
You need to use multipe async blocks to run things concurrently

Here is something else that might help you

https://github.com/socketry/benchmark-http/blob/master/lib/benchmark/http/command/spider.rb

it's a spider. It's limited to 4 connections but will fetch things as fast as possible.

If you want a map-reduce style appraoch, you can put the results into an `Async::Queue` and read them out in another task. The sky is the limit. Feel free to post your code here for further discussion.

**Comment by ioquatix** [comment]
I've updated the example to show you how to fetch multiple comics at the same time.

**Comment by vfonic** [comment]
I'm still trying things out.

This seems to take ~15s:

```ruby
require 'nokogiri'
require 'pry'
require 'async'
require 'async/http/internet'
require 'async/rest'
require 'async/rest/wrapper/url_encoded'

module XKCD
  module Wrapper
    # This defines how we interact with the XKCD service.
    class HTML < Async::REST::Wrapper::URLEncoded
      TEXT_HTML = "text/html"

      # How to process the response body.
      class Parser < ::Protocol::HTTP::Body::Wrapper
        def join
          Nokogiri::HTML(super)
        end
      end

      # We wrap the response body with the parser (it could incrementally parse the body).
      def wrap_response(response)
        if body = response.body
          response.body = Parser.new(body)
        end
      end

      def process_response(request, response)
        if content_type = response.headers['content-type']
          if content_type.start_with? TEXT_HTML
            wrap_response(response)
          else
            raise Error, "Unknown content type: #{content_type}!"
          end
        end

        return response
      end
    end
  end

  # A comic representation.
  class Comic < Async::REST::Representation[Wrapper::HTML]
    def image_url
      self.value.css("#comic img").attribute("src").text
    end
  end
end

urls = (2155..2205).to_a.map{|i| "https://xkcd.com/#{i}/" }
urls.each do |url|
  Async do
    Async::REST::Resource.for(url) do |resource|
      representation = resource.get(XKCD::Comic)

      p representation.value.css('#ctitle').first.text
    end
  end
end
```

While this takes (on the second run (?)) about ~5s:

```ruby
require 'nokogiri'
require 'pry'
require 'async'
require 'async/http/internet'
require 'async/rest'
require 'async/rest/wrapper/url_encoded'

urls = (2155..2205).to_a.map{|i| "https://xkcd.com/#{i}/" }
Async do
  internet = Async::HTTP::Internet.new
  urls.each do |url|
    response = internet.get(url)
    page = Nokogiri::HTML(response.read)
    p page.css('#ctitle').first.text
  end
ensure
  internet.close
end
```

What am I missing? (using async-rest 0.11.0 from github)
How's it that `async-http` sometimes runs so slow on the first run? How can I ensure this doesn't happen in production?

EDIT: Great, let me try the new `async-rest` example. :)

**Comment by ioquatix** [comment]
The problem is you need to share the underlying connection.

`Async::REST::Resource.for(url) do |resource|` for each request makes a new connection cannot reuse connection.

Same with `Async::Internet.new`. You must share it between all requests for optimal reuse/keep-alive, etc.

**Comment by vfonic** [comment]
Ahhhh! I have no idea how I can do that with `Async::HTTP::Internet.new`, but this is superfast:

:tada:

```ruby
URL = 'https://xkcd.com/'
Async::REST::Resource.for(URL) do |resource|
  (2155..2205).each do |id|
    Async do
      representation = resource.with(path: "/#{id}/").get(XKCD::Comic)

      p representation.value.css('#ctitle').first.text
    end
  end
end
```

**Comment by ioquatix** [comment]
```ruby
#!/usr/bin/env ruby

require 'nokogiri'
require 'pry'
require 'async'
require 'async/http/internet'
require 'async/rest'
require 'async/rest/wrapper/url_encoded'

urls = (2155..2205).to_a.map{|i| "https://xkcd.com/#{i}/" }
Async do
	internet = Async::HTTP::Internet.new
	titles = urls.map do |url|
		Async do
			response = internet.get(url)
			page = Nokogiri::HTML(response.read)
			[url, page.css('#ctitle').first.text]
		end
	end.map(&:wait)
	
	pp titles
ensure
	internet.close
end
```

**Comment by ioquatix** [comment]
We also thinking of monkey patching in `async_map`, `async_each` and so on. It would simplify the above code greatly.

**Comment by vfonic** [comment]
I see. Was it that `internet.close` was being called on every `.each`?

What does `.wait` do? I don't see much of difference running with or without `.map(&:wait)`. It seems a bit faster, but can't say.

Thank you for answering all the questions so far! 🙏 👍

**Comment by ioquatix** [comment]
The url loop was doing it one at a time, I just put an async block inside it so they could work concurrently.

`.wait` waits for the task (and therefore response) to finish.

So, we map all urls to tasks which are fetching the HTML, then once we've made all the tasks, we start calling .wait on them, which reutrns the result `[url, page.css(...)]` in a convenient array (one per task).

**Comment by vfonic** [comment]
I think I understand it all now. I caught your code example before you were using/assigning the result of .map, so I was confused by calling wait.

The only reason why we need the outer Async is so that we can instantiate and use `Async::HTTP::Internet`, right?

**Comment by ioquatix** [comment]
> The only reason why we need the outer Async is so that we can instantiate and use Async::HTTP::Internet, right?

Yes. Because HTTP/2 requires a background reader.
